src/Polysemy/Internal/Union.hs

```diff
@@ -56,77 +59,79 @@ import Polysemy.Internal.CustomErrors
-         -- can't expose that fact; such a thing would prevent 'Polysemy.Sem'
-         -- from being a 'Monad'.
-       , weaveInspect :: forall x. f x -> Maybe x
-         -- ^ A function for attempting to see inside an @f@. This is no
-         -- guarantees that such a thing will succeed (for example,
-         -- 'Polysemy.Error.Error' might have 'Polysemy.Error.throw'n.)
-       }
-    -> Weaving e n b
+    :: forall f e rBefore a resultType mAfter. (Functor f)
+    => !(WeavingDetails f e rBefore a resultType mAfter)
+    -> Weaving e mAfter resultType
+
+data Functor f => WeavingDetails f e rBefore a resultType mAfter =
+  WeavingDetails {
+    weaveEffect :: e (Sem rBefore) a
+    -- ^ The original effect GADT originally lifted via
+    -- 'Polysemy.Internal.send'.
+  , weaveState :: f ()
+    -- ^ A piece of state that other effects' interpreters have already
+    -- woven through this 'Weaving'. @f@ is a 'Functor', so you can always
+    -- 'fmap' into this thing.
+  , weaveDistrib :: forall x. f (Sem rBefore x) -> mAfter (f x)
+    -- ^ Distribute @f@ by transforming @Sem rBefore@ into @mAfter@. We have an
+    -- invariants on @rBefore@, which means in actuality this function looks like
+    -- @f ('Polysemy.Sem' rBefore x) -> mAfter (f x)@.
+  , weaveResult :: f a -> resultType
+    -- ^ Even though @f a@ is the moral resulting type of 'Weaving', we
+    -- can't expose that fact; such a thing would prevent 'Polysemy.Sem'
+    -- from being a 'Monad'.
+  , weaveInspect :: forall x. f x -> Maybe x
+    -- ^ A function for attempting to see inside an @f@. This is no
+    -- guarantees that such a thing will succeed (for example,
+    -- 'Polysemy.Error.Error' might have 'Polysemy.Error.throw'n.)
+  }
 
 instance Functor (Weaving e m) where
-  fmap f (Weaving e s d f' v) = Weaving e s d (f . f') v
+  fmap f (Weaving (WeavingDetails e s d f' v)) =
+    Weaving $ WeavingDetails e s d (f . f') v
   {-# INLINE fmap #-}
 
 
 
 weave
-    :: (Functor s, Functor m, Functor n)
+    :: (Functor s, Functor mMid, Functor mAfter)
     => s ()
-    -> (∀ x. s (m x) -> n (s x))
+    -> (∀ x. s (mMid x) -> mAfter (s x))
     -> (∀ x. s x -> Maybe x)
-    -> Union r m a
-    -> Union r n (s a)
-weave s' d v' (Union w (Weaving e s nt f v)) = Union w $
-    Weaving e (Compose $ s <$ s')
+    -> Union r mMid a
+    -> Union r mAfter (s a)
+weave s' d v' (Union (UnionDetails w (Weaving (WeavingDetails e s nt f v)))) =
+  Union $ UnionDetails w $ Weaving $ WeavingDetails
+              e (Compose $ s <$ s')
               (fmap Compose . d . fmap nt . getCompose)
               (fmap f . getCompose)
               (v <=< v' . getCompose)
 {-# INLINE weave #-}
 
 
 hoist
-    :: ( Functor m
-       , Functor n
-       )
-    => (∀ x. m x -> n x)
-    -> Union r m a
-    -> Union r n a
-hoist f' (Union w (Weaving e s nt f v)) = Union w $ Weaving e s (f' . nt) f v
+    :: (∀ x. mBefore x -> mAfter x)
+    -> Union r mBefore a
+    -> Union r mAfter a
```

Comment:
Same as above: I don't find `mBefore` and `mAfter` helpful here. They don't convey any extra information.